### ISSUE TITLE
Medic Gateway runs into document update conflicts while trying to upload sms status

### ISSUE THREAD

**Issue by henokgetachew** [medic/medic-webapp#3443]
Instance to test this with: clone-kilifikids-kaiti.
What this example below causes for example is that the sms is reported as `Pending` when it has already been picked up by gateway and it has been trying to communicate that the status is now `forwarded-to-gateway`.

Logs from API:
```
2017-05-04T11:33:00.400Z - error: WARN sms-gateway { Error: Document update conflict.                                                                                                                                                         
    at Request._callback (/srv/storage/gardener/data/working_dir/aHR0cDovL2xvY2FsaG9zdDo1OTg0L21lZGljL19kZXNpZ24vbWVkaWMvbWVkaWMtYXBp/node_modules/medic-api/node_modules/nano/lib/nano.js:248:15)                                            
    at Request.self.callback (/srv/storage/gardener/data/working_dir/aHR0cDovL2xvY2FsaG9zdDo1OTg0L21lZGljL19kZXNpZ24vbWVkaWMvbWVkaWMtYXBp/node_modules/medic-api/node_modules/request/request.js:188:22)                                      
    at emitTwo (events.js:106:13)                                                                                                                                                                                                             
    at Request.emit (events.js:191:7)                                                                                                                                                                                                         
    at Request.<anonymous> (/srv/storage/gardener/data/working_dir/aHR0cDovL2xvY2FsaG9zdDo1OTg0L21lZGljL19kZXNpZ24vbWVkaWMvbWVkaWMtYXBp/node_modules/medic-api/node_modules/request/request.js:1171:10)                                       
    at emitOne (events.js:96:13)                                                                                                                                                                                                              
    at Request.emit (events.js:188:7)                                                                                                                                                                                                         
    at IncomingMessage.<anonymous> (/srv/storage/gardener/data/working_dir/aHR0cDovL2xvY2FsaG9zdDo1OTg0L21lZGljL19kZXNpZ24vbWVkaWMvbWVkaWMtYXBp/node_modules/medic-api/node_modules/request/request.js:1091:12)                               
    at IncomingMessage.g (events.js:291:16)                                                                                                                                                                                                   
  name: 'Error',                                                                                                                                                                                                                              
  error: 'conflict',                                                                                                                                                                                                                          
  reason: 'Document update conflict.',                                                                                                                                                                                                        
  scope: 'couch',                                                                                                                                                                                                                             
  statusCode: 409,                                                                                                                                                                                                                            
  request:                                                                                                                                                                                                                                    
   { method: 'PUT',                                                                                                                                                                                                                           
     headers:                                                                                                                                                                                                                                 
      { 'content-type': 'application/json',                                                                                                                                                                                                   
        accept: 'application/json' },                                                                                                                                                                                                         
     uri: 'http://XXXXXX:XXXXXX@localhost:5984/medic/_design/medic/_update/update_message_task/00f7e3944e5d017cf91f172047cb460d',                                                                                                             
     body: '{"state":"forwarded-to-gateway","message_id":"d5d8c6d5-9d05-412a-8d4d-b03e51f04fe9"}' },                                                                                                                                          
  headers:                                                                                                                                                                                                                                    
   { date: 'Thu, 04 May 2017 11:33:00 GMT',                                                                                                                                                                                                   
     'content-type': 'application/json',                                                                                                                                                                                                      
     'cache-control': 'must-revalidate',                                                                                                                                                                                                      
     statusCode: 409,                                                                                                                                                                                                                         
     uri: 'http://XXXXXX:XXXXXX@localhost:5984/medic/_design/medic/_update/update_message_task/00f7e3944e5d017cf91f172047cb460d' },                                                                                                           
  errid: 'non_200',                                                                                                                                                                                                                           
  description: 'couch returned 409' }
```

**Comment by sglangevin** [comment]
This is likely a blocker for upgrades, so I've added it to the current milestone to see if someone can take a look.

**Comment by sglangevin** [comment]
@henokgetachew what version of webapp was running?

**Comment by sglangevin** [comment]
Looks like 2.11.0. Can you update to v2.11.1-beta.2 and re-test?

**Comment by henokgetachew** [comment]
Since this instance is on alpha, it's now been updated to the latest alpha - 2.12.0-alpha.5680.  @christinewere can you retest medic gateway related issues again? Sending and receiving messages to and from the instance and checking whether the statuses are reflected on the instance correctly?

**Comment by SCdF** [comment]
@henokgetachew could you help with specific reproduction steps, and ideally a link (private on slack if you like) of a instance where this is occurring, with the document that is having problems?

**Comment by SCdF** [comment]
So:
![image](https://cloud.githubusercontent.com/assets/583851/25805454/9ccc0668-33f7-11e7-9429-59e6b15c7e78.png)
![image](https://cloud.githubusercontent.com/assets/583851/25805470/affef420-33f7-11e7-9835-f139049cc387.png)
And:
![image](https://cloud.githubusercontent.com/assets/583851/25805440/8ff9fda0-33f7-11e7-84e8-66f9b84aebf3.png)

Make me think this isn't because of sentinel, since no sentinel transitions have written to this document (apparently, though it's not 100% convincing) in a long time.

**Comment by SCdF** [comment]
I have no idea how this can be a 409 actually, because you're not passing a rev, so how can it conflict? It implies that in the time it takes the _update code to run something else has updated the document. Which could happen once, but to happen over and over again… suspiciously weird.

Dumb question, but is there a chance there are two instances of whomever is calling this function (sentinel / gateway, I'm not quite clear yet)?

**Comment by SCdF** [comment]
![image](https://cloud.githubusercontent.com/assets/583851/25812460/4706028a-340e-11e7-8826-e2ed1310c3c5.png)
Right then.

**Comment by SCdF** [comment]
OK. Where I am at with this there are presumably a lot of messages that need to be sent to gateway that originate from the same report. I am wondering if somehow we are actually running these requests asynchronously (even though the code looks fine after a reasonable walkthrough of it) and at least two of them are basically trying to update their same parent document at the same time.

**Comment by SCdF** [comment]
FWIW I can replicate this if two requests to the medic-api gateway run at the same time: 
```sh
curl http://admin:pass@localhost:5988/api/sms -d '{}' & curl http://admin:pass@localhost:5988/api/sms -d '{}' 
```

But if I run them one after other it works for me fine locally.

So I'm not really sure why this is failing. I'm going to take a look at the audit log, but I think maybe the best I can do is harden this code to make it less susceptible to these sorts of problems, and have it attempt to resolve 409s itself instead of throwing them back to gateway.

**Comment by sglangevin** [comment]
Thanks for looking into this, @SCdF. @christinewere have you tested now that the instance was upgraded? Sounds like hardening the code is a good solution either way.

**Comment by alxndrsn** [comment]
@SCdF It sounds like gateway could do with stronger protection against running requests in parallel.  Right now it shouldn't be likely, but definitely possible - especially if api's response times are slow.

**Comment by SCdF** [comment]
@alxndrsn yeah totally. I know it sounds horrid, but I'm going to seriously look into some form of mutex for that call, since it a) has no reason being run by more than one thing at once, and b) does X*Y*Z things that will conflict with multieple requestors which we cannot collapse into a single couchdb request.

**Comment by rosteve** [comment]
The messages are now showing  `forwarded` but they are being forwarded to the web app in duplicates .. logs for the gateway is attached
[logs.txt](https://github.com/medic/medic-webapp/files/989357/logs.txt)

**Comment by christinewere** [comment]
Thank you @rosteve

**Comment by SCdF** [comment]
Went to check the logs to see if we could definitively see if someone was accessing at the same time:
 - nginx logs don't indicate that anyone is
 - api logs have been surprise deleted by gardener already, so dead end there.

I'm going to look into refactoring the code

**Comment by SCdF** [comment]
[Work complete](https://www.youtube.com/watch?v=bupagiROLV8).

PRs:
 - https://github.com/medic/medic-webapp/pull/3475/files
 - https://github.com/medic/medic-api/pull/152/files

Changes:
 - Pulled out updating tasks via message state changes from CouchDB and into API. This lets us do more than one at a time
 - Refactored the code that dealt with this to use bulk save instead of saving things one by one
 - Checks the result of the bulk save and takes any state changes that fails and re-applies them from fresh DB assets (finding which docs are relevant for which state changes again etc). It does this three times before giving up and returning an error
 - The api follows the rest of medic-api's error reporting, which is not the best, but at least consistent. If anything goes wrong you get a 500 and `server error` returned. We could improve this across the board in the future.

**Comment by SCdF** [comment]
Hey @alxndrsn please review the above two PRs. Let me know if you have any questions.

**Comment by alxndrsn** [comment]
Is there any update to the external `/messages` api after refactoring of message controller and message utils?  https://github.com/medic/medic-api#messages

**Comment by SCdF** [comment]
@alxndrsn there should be no api changes anywhere in this code, no. With the exception that we no longer hide errors, but the documentation never stated we did that.

**Comment by SCdF** [comment]
Merged to master.

**Comment by sglangevin** [comment]
@henokgetachew there is a beta release out with this fix in it - v2.12.0-beta.2. Can you upgrade the affected instance to this beta and do acceptance testing so we can release 2.12.0 final?

**Comment by sglangevin** [comment]
Tested on v2.12.0-beta.5 on clone-kilifikids-kaiti.app and is working well. Moving to ready.
